File: pillars/bfm/src/contract/account.ts
import { z } from 'zod';

/**
 * What the account holds today, as the phone reads it — finance's own
 * `AccountBalance` (POPS-2880) minus `anchor`: the phone shows a date and an
 * `inconsistent` flag, not a checkpoint id to link to. Ledger-signed like
 * everywhere else in this contract — positive is money held, negative is
 * money owed, for assets and liabilities alike; nothing here negates it.
 *
 * `basis: 'transactions'` means finance found no checkpoint to anchor on and
 * the figure is the sum of whatever was imported — net flow, not a balance.
 * A caller that renders this the same way it renders `'checkpoint'` is
 * showing a number that can drift arbitrarily far from what the account
 * actually holds; POPS-2848 owns making that distinction visible.
 */
export const MobileAccountBalanceSchema = z.object({
  balanceCents: z.number().int(),
  asOf: z.string(),
  basis: z.enum(['checkpoint', 'transactions']),
  inconsistent: z.boolean(),
});

export type MobileAccountBalance = z.infer<typeof MobileAccountBalanceSchema>;

/**
 * One account, as the phone reads it.
 *
 * `kind` is left an open string for the same reason as
 * {@link MobileTransactionSchema.shape.type}: finance adding an account kind
 * must not fail every account already on a phone to decode.
 */
export const MobileAccountSchema = z.object({
  id: z.string(),
  name: z.string(),
  kind: z.string(),
  /** ISO 4217 code, or a points-program code — finance's own vocabulary (POPS-2802). */
  currency: z.string(),
  archived: z.boolean(),
  /** `institutions` id this account is held at, or `null` for cash and person accounts. */
  institutionId: z.string().nullable(),
  balance: MobileAccountBalanceSchema,
});

export type MobileAccount = z.infer<typeof MobileAccountSchema>;

/**
 * Every account the list screen shows.
 *
 * Unpaged, unlike {@link MobileTransactionsPageSchema}: an account list is
 * small enough in practice to send whole (`AccountsRepository.accounts()`'s
 * own reasoning), so there is no cursor to get wrong.
 */
export const MobileAccountsPageSchema = z.object({
  data: z.array(MobileAccountSchema),
});

export type MobileAccountsPage = z.infer<typeof MobileAccountsPageSchema>;

File: clients/ios/Packages/AppCore/Sources/AppCore/Accounts/Account.swift
import Foundation

/// What a balance is anchored on, mirroring finance's own `basis` (ADR-051).
///
/// The distinction is not cosmetic. ``checkpoint`` means the figure is pinned
/// to a balance somebody read off the account and adjusted by the transactions
/// since; ``transactions`` means finance found no checkpoint at all and the
/// figure is the sum of whatever happens to have been imported — net flow,
/// which can sit arbitrarily far from what the account actually holds. A
/// screen that renders the two identically is presenting a guess as a fact.
public enum BalanceBasis: Hashable, Sendable {
    case checkpoint
    case transactions
}

/// One account, in the app's own vocabulary rather than the wire's.
///
/// Mirrors ``Transaction``'s reason for existing as its own type: a feature is
/// written against this and never against a generated one, so the contract can
/// move without a feature noticing.
public struct Account: Hashable, Sendable, Identifiable {
    public let id: String
    public let name: String
    public let kind: AccountKind
    /// Minor units, signed in the account's own terms — a credit card's
    /// positive balance is money owed, a person ledger's positive balance is
    /// owed to you. The currency travels with the amount.
    public let balance: MoneyAmount
    public let archived: Bool
    /// The institution this account is held at, when it has one. Cash and
    /// person ledgers have none.
    public let institutionName: String?
    /// Who a gift card or a person ledger is with.
    public let contact: String?
    /// The date the balance is true as of — for a ``BalanceBasis/checkpoint``
    /// figure, when it was last anchored; for a ``BalanceBasis/transactions``
    /// one, simply how far the ledger runs.
    public let balanceAsOf: Date?
    /// What the balance is anchored on, which decides how honestly it can be
    /// phrased. See ``BalanceBasis``.
    public let balanceBasis: BalanceBasis
    /// Finance's own reading that the ledger and a checkpoint disagree by more
    /// than rounding. Always `false` under ``BalanceBasis/transactions``,
    /// where there is no checkpoint to disagree with.
    ///
    /// Carried but not yet drawn: how a disagreement reads on a phone is a
    /// design question rather than a rendering one (POPS-2927).
    public let balanceInconsistent: Bool
    /// Gift cards only.
    public let expiresOn: Date?
    /// How many transactions this account holds — every row finance has for
    /// it, pending and transfer rows included (POPS-2924).
    ///
    /// Optional so a repository or fixture that has none to report can leave
    /// it out rather than inventing a number; a screen drops the clause
    /// rather than showing a zero that reads as "no transactions". The BFM
    /// wire itself always carries a real count.
    public let transactionCount: Int?

    public init(
        id: String,
        name: String,
        kind: AccountKind,
        balance: MoneyAmount,
        archived: Bool,
        institutionName: String? = nil,
        contact: String? = nil,
        balanceAsOf: Date? = nil,
        balanceBasis: BalanceBasis = .transactions,
        balanceInconsistent: Bool = false,
        expiresOn: Date? = nil,
        transactionCount: Int? = nil
    ) {
        self.id = id
        self.name = name
        self.kind = kind
        self.balance = balance
        self.archived = archived
        self.institutionName = institutionName
        self.contact = contact
        self.balanceAsOf = balanceAsOf
        self.balanceBasis = balanceBasis
        self.balanceInconsistent = balanceInconsistent
        self.expiresOn = expiresOn
        self.transactionCount = transactionCount
    }
}
